mountpoint-s3-client/src/imds_crt_client.rs

```diff
@@ -0,0 +1,114 @@
+use std::time::Duration;
+
+use futures::channel::oneshot;
+use futures::Future;
+use mountpoint_s3_crt::auth::imds_client::ImdsClient;
+use mountpoint_s3_crt::auth::imds_client::ImdsClientConfig;
+use mountpoint_s3_crt::common::allocator::Allocator;
+use mountpoint_s3_crt::common::error;
+use mountpoint_s3_crt::io::channel_bootstrap::{ClientBootstrap, ClientBootstrapOptions};
+use mountpoint_s3_crt::io::event_loop::EventLoopGroup;
+use mountpoint_s3_crt::io::host_resolver::{HostResolver, HostResolverDefaultOptions};
+use mountpoint_s3_crt::io::retry_strategy::ExponentialBackoffJitterMode;
+use mountpoint_s3_crt::io::retry_strategy::{RetryStrategy, StandardRetryOptions};
+use pin_project::pin_project;
+use thiserror::Error;
+
+#[derive(Debug)]
+/// Instance Metadata Service (IMDS) client responsible for sending ec2 instance metadata query requests.
+pub struct ImdsCrtClient {
+    imds_client: ImdsClient,
+    event_loop_group: EventLoopGroup,
+    allocator: Allocator,
+}
+
+impl ImdsCrtClient {
+    pub fn new() -> Result<Self, error::Error> {
+        let allocator = Allocator::default();
+
+        let mut event_loop_group = EventLoopGroup::new_default(&allocator, None, || {}).unwrap();
+
+        let resolver_options = HostResolverDefaultOptions {
+            max_entries: 8,
+            event_loop_group: &mut event_loop_group,
+        };
+
+        let mut host_resolver = HostResolver::new_default(&allocator, &resolver_options).unwrap();
+
+        let bootstrap_options = ClientBootstrapOptions {
+            event_loop_group: &mut event_loop_group,
+            host_resolver: &mut host_resolver,
+        };
+
+        let client_bootstrap = ClientBootstrap::new(&allocator, &bootstrap_options).unwrap();
+
+        let mut retry_strategy_options = StandardRetryOptions::default(&mut event_loop_group);
+
+        // Match the Java SDK retry strategies:
+        // 1.max retries: https://tinyurl.com/3vbn5ty6
+        // 2.backoff strategy: https://tinyurl.com/ycmta4vw
+        // 3.backoff duration: https://tinyurl.com/4rfzsdva
+        retry_strategy_options.backoff_retry_options.max_retries = 3;
+        retry_strategy_options.backoff_retry_options.backoff_scale_factor = Duration::from_millis(20000);
+        retry_strategy_options.backoff_retry_options.jitter_mode = ExponentialBackoffJitterMode::Full;
+        let retry_strategy = RetryStrategy::standard(&allocator, &retry_strategy_options).unwrap();
+
+        let mut client_config = ImdsClientConfig::new();
+
+        client_config
+            .client_bootstrap(client_bootstrap)
+            .retry_strategy(retry_strategy);
+
+        let imds_client = ImdsClient::new(&allocator, client_config).unwrap();
+
+        Ok(Self {
+            imds_client,
+            event_loop_group,
+            allocator,
+        })
+    }
+
+    /// Query the instnace type mountpoint-s3 running on.
+    pub fn make_instance_type_query(&self) -> Result<ImdsQueryRequest, ImdsQueryRequestError> {
```

Comment:
We haven't put thoughts to Lambda use case yet.

Reply:
We need to at least test this code somewhere that isn't an EC2 instance.

Reply:
Tested on my local laptop(ubuntu), the program can run as normal and it will shows the following errors when calling  `retrieve_instance_type`
```
2023-04-03T19:57:45.731861Z  WARN awscrt::Unknown: id=0x55f099dfa390: IMDS Client failed to acquire a connection, error code 1048(socket operation timed out.)    
2023-04-03T19:57:45.731911Z ERROR awscrt::Unknown: (id=0x55f099dfa390) IMDS client failed to update the token from IMDS. 
2023-04-03T19:57:48.732022Z  WARN awscrt::Unknown: id=0x55f099dfa390: IMDS Client failed to acquire a connection, error code 1048(socket operation timed out.)
```

Reply:
One concern here - we've added a delay during mount whenever we run on something that isn't an EC2 instance.

Customers can make it go away by specifying a value for throughput, but maybe we can be smarter and fail earlier. Maybe a file exists on an EC2 instance that we can check before making the IMDS calls?